src/Sarif.Converters/FortifyFprConverter.cs

```diff
@@ -0,0 +1,631 @@
+                    {
+                        ParseVulnerabilities();
+                    }
+                    else if (AtStartOfNonEmpty(_strings.Description))
+                    {
+                        ParseRuleFromDescription();
+                    }
+                    else if (AtStartOfNonEmpty(_strings.Snippets))
+                    {
+                        ParseSnippets();
+                    }
+                    else if (AtStartOfNonEmpty(_strings.CommandLine))
+                    {
+                        ParseCommandLine();
+                    }
+                    else if (AtStartOfNonEmpty(_strings.Errors))
+                    {
+                        ParseErrors();
+                    }
+                    else if (AtStartOfNonEmpty(_strings.MachineInfo))
+                    {
+                        ParseMachineInfo();
+                    }
+                }
+            }
+        }
+
+        private void ParseUuid()
+        {
+            _runId = _reader.ReadElementContentAsString();
+        }
+
+        private void ParseCreatedTimestamp()
+        {
+            string date = _reader.GetAttribute(_strings.DateAttribute);
+            string time = _reader.GetAttribute(_strings.TimeAttribute);
+            if (!string.IsNullOrEmpty(date) && !string.IsNullOrEmpty(time))
+            {
+                string dateTime = date + "T" + time;
+                _invocation.StartTime = DateTime.Parse(dateTime, CultureInfo.InvariantCulture);
+            }
+
+            // Step past the empty element.
+            _reader.Read();
+        }
+
+        private void ParseBuild()
+        {
+            _reader.Read();
+            while (!AtEndOf(_strings.Build))
+            {
+                if (AtStartOfNonEmpty(_strings.BuildId))
+                {
+                    _automationId = _reader.ReadElementContentAsString();
+                }
+                else if (AtStartOfNonEmpty(_strings.SourceFiles))
+                {
+                    ParseSourceFiles();
+                }
+                else
+                {
+                    _reader.Read();
+                }
+            }
+        }
+
+        private void ParseSourceFiles()
+        {
+            _reader.Read();
+            while (!AtEndOf(_strings.SourceFiles))
+            {
+                if (AtStartOfNonEmpty(_strings.File))
+                {
+                    ParseFile();
+                }
+                else
+                {
+                    _reader.Read();
+                }
+            }
+        }
+
+        private void ParseFile()
+        {
+            string fileType = _reader.GetAttribute(_strings.TypeAttribute);
+            int length = 0;
+            string sizeAttribute = _reader.GetAttribute(_strings.SizeAttribute);
+            if (sizeAttribute != null)
+            {
+                if (!int.TryParse(sizeAttribute, out length))
+                {
+                    length = 0;
```

Comment:
Re: error reporting: I'll look at what other converters do when the input file is invalid. One possibility is to emit toolNotifications. But then the toolNotifications would be a mix of notifications from the original scanner (the Errors/Error elements in the FPR file) and from the converter.

---
In reply to: [94534782](https://github.com/Microsoft/sarif-sdk/pull/554#discussion_r94534782) [](ancestors = 94534782,94534761)